src/Facing.php

```diff
@@ -56,120 +38,112 @@ private function __construct(){
+			self::DOWN, self::UP => Axis::Y,
+			self::NORTH, self::SOUTH => Axis::Z,
+			self::WEST, self::EAST => Axis::X,
+		};
+	}
+
+	/**
+	 * @phpstan-return array{-1|0|1, -1|0|1, -1|0|1}
+	 */
+	public function offset() : array{
+		return match($this){
+			self::DOWN  => [ 0, -1,  0],
+			self::UP    => [ 0, +1,  0],
+			self::NORTH => [ 0,  0, -1],
+			self::SOUTH => [ 0,  0, +1],
+			self::WEST  => [-1,  0,  0],
+			self::EAST  => [+1,  0,  0]
+		};
 	}
 
 	/**
 	 * Returns whether the direction is facing the positive of its axis.
 	 */
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public function isPositive() : bool{
+		return $this === self::UP || $this === self::SOUTH || $this === self::EAST;
 	}
 
 	/**
 	 * Returns the opposite Facing of the specified one.
-	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
 	 */
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public function opposite() : Facing{
+		return match($this){
+			self::DOWN => self::UP,
+			self::UP => self::DOWN,
+			self::NORTH => self::SOUTH,
+			self::SOUTH => self::NORTH,
+			self::WEST => self::EAST,
+			self::EAST => self::WEST,
+		};
 	}
 
 	/**
 	 * Rotates the given direction around the axis.
 	 *
-	 * @throws \InvalidArgumentException if not possible to rotate $direction around $axis
+	 * @throws \InvalidArgumentException if not possible to rotate this direction around $axis
 	 */
-	public static function rotate(int $direction, int $axis, bool $clockwise) : int{
-		if(!isset(self::CLOCKWISE[$axis])){
-			throw new \InvalidArgumentException("Invalid axis $axis");
-		}
-		if(!isset(self::CLOCKWISE[$axis][$direction])){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public function rotate(Axis $axis, bool $clockwise) : Facing{
+		$rotated = match($axis){
+			Axis::Y => match($this){
+				self::NORTH => self::EAST,
+				self::EAST => self::SOUTH,
+				self::SOUTH => self::WEST,
+				self::WEST => self::NORTH,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			},
+			Axis::Z => match($this){
+				self::UP => self::EAST,
+				self::EAST => self::DOWN,
+				self::DOWN => self::WEST,
+				self::WEST => self::UP,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			},
+			Axis::X => match($this){
+				self::UP => self::NORTH,
+				self::NORTH => self::DOWN,
+				self::DOWN => self::SOUTH,
+				self::SOUTH => self::UP,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			}
```

Comment:
The default cases could just fall through to `null`, then we could just check if `$rotated` is `null` after the match so as not to duplicate exception messages 3 times